File: example_usage.py
#!/usr/bin/env python3
"""
Biometric Device Middleware - Example Usage

This script demonstrates how to use the BiometricDeviceMiddleware to interact
with biometric devices using the specified protocol.
"""

import time
from datetime import datetime
from biometric_middleware import BiometricDeviceMiddleware, BiometricDeviceError


def main():
    # Initialize the middleware
    device_ip = "192.168.1.100"  # Replace with your device IP
    api_key = "your_api_key"     # Replace with your API key (if configured)
    
    # Create middleware instance
    device = BiometricDeviceMiddleware(device_ip, api_key=api_key)
    
    print("=== Biometric Device Middleware Demo ===\n")
    
    try:
        # 1. Get device information
        print("1. Getting device information...")
        device_info = device.get_version_info()
        print(f"   Firmware version: {device_info.get('firmware_version')}")
        print(f"   Face algorithm: {device_info.get('face_algorithm_version', 'N/A')}")
        print(f"   Fingerprint algorithm: {device_info.get('fp_algorithm_version', 'N/A')}")
        
        # Get device capabilities
        capabilities = device.get_device_capabilities()
        print(f"   Capabilities: {capabilities}")
        
        # Get device usage
        usage = device.get_current_usage()
        print(f"   Current usage: {usage}")
        
        print()
        
        # 2. Time synchronization
        print("2. Time synchronization...")
        current_time = device.get_device_time()
        print(f"   Current device time: {current_time}")
        
        # Set device time to current system time
        device.set_device_time()
        new_time = device.get_device_time()
        print(f"   Updated device time: {new_time}")
        print()
        
        # 3. User management
        print("3. User management...")
        
        # Get all user IDs
        user_ids = device.get_all_user_ids()
        print(f"   Found {len(user_ids)} users: {user_ids[:10]}")  # Show first 10
        
        # Create a test user
        test_user_id = "999"
        print(f"   Creating test user with ID: {test_user_id}")
        device.set_user_info(
            user_id=test_user_id,
            name="Test User",
            department="Engineering",
            privilege="user",
            password="12345"
        )
        
        # Get user information
        user_info = device.get_user_info(test_user_id)
        print(f"   Test user info: {user_info}")
        print()
        
        # 4. Device settings
        print("4. Device settings...")
        
        # Get and display current volume
        volume = device.get_sound_volume()
        print(f"   Current sound volume: {volume}")
        
        # Get verify mode
        verify_mode = device.get_verify_mode()
        print(f"   Current verify mode: {verify_mode}")
        
        print()
        
        # 5. Biometric enrollment demo (face)
        print("5. Biometric enrollment demo...")
        
        # Lock device before enrollment to prevent interference
        print("   Locking device for enrollment...")
        device.lock_device(True)
        
        try:
            # Start face enrollment
            print("   Starting face enrollment...")
            print("   Please look at the device camera when prompted...")
            
            job_id = device.begin_enroll_face()
            print(f"   Enrollment job started with ID: {job_id}")
            
            # Wait for enrollment completion (with timeout)
            print("   Waiting for enrollment completion...")
            try:
                result = device.wait_for_enrollment_completion(job_id, timeout=30)
                print(f"   Enrollment successful! Face data length: {len(result.get('face_data', ''))}")
                
                # Update user with face data
                device.set_user_info(test_user_id, face_data=result.get('face_data'))
                print("   User updated with face data")
                
            except BiometricDeviceError as e:
                print(f"   Enrollment failed or timed out: {e}")
                
        finally:
            # Always unlock device
            device.lock_device(False)
            print("   Device unlocked")
        
        print()
        
        # 6. Attendance log management
        print("6. Attendance log management...")
        
        # Get recent attendance logs
        logs = device.get_attend_log()
        print(f"   Retrieved {len(logs.get('logs', []))} attendance records")
        
        # Show first few logs
        for i, log in enumerate(logs.get('logs', [])[:3]):
            print(f"   Log {i+1}: User {log.get('user_id')} at {log.get('time')} via {log.get('mode')}")
        
        print()
        
        # 7. Network configuration info
        print("7. Network configuration...")
        network_config = device.get_network_config()
        
        if 'ethernet' in network_config:
            eth_config = network_config['ethernet']
            running = eth_config.get('running', {})
            print(f"   Ethernet IP: {running.get('address')}")
            
        if 'wlan' in network_config:
            wlan_config = network_config['wlan']
            running = wlan_config.get('running', {})
            print(f"   WiFi IP: {running.get('address')}")
        
        print()
        
        # 8. Clean up test user
        print("8. Cleaning up...")
        try:
            device.delete_user_info(test_user_id)
            print(f"   Test user {test_user_id} deleted")
        except BiometricDeviceError:
            print(f"   Test user {test_user_id} not found (already deleted?)")
            
    except BiometricDeviceError as e:
        print(f"Error communicating with device: {e}")
        if hasattr(e, 'error_code'):
            print(f"Error code: {e.error_code}")
    except Exception as e:
        print(f"Unexpected error: {e}")
    
    print("\n=== Demo completed ===")


def demo_photo_to_face():
    """
    Demonstrate photo to face data conversion
    """
    print("\n=== Photo to Face Data Demo ===")
    
    device_ip = "192.168.1.100"  # Replace with your device IP
    device = BiometricDeviceMiddleware(device_ip)
    
    # This is a placeholder - you would load an actual base64-encoded JPG image
    # photo_base64 = load_photo_as_base64("path/to/photo.jpg")
    
    print("Note: This demo requires a base64-encoded JPG photo.")
    print("To use this feature, uncomment and modify the code above.")
    
    # Example usage:
    # try:
    #     result = device.photo_to_face_data(photo_base64)
    #     if result.get('state') == 'succeeded':
    #         face_data = result.get('face_data')
    #         print(f"Face data extracted successfully! Length: {len(face_data)}")
    #         # You can now use this face_data with set_user_info
    #     else:
    #         print(f"Photo conversion failed: {result.get('state')}")
    # except BiometricDeviceError as e:
    #     print(f"Error: {e}")


def demo_attendance_uploader():
    """
    Demonstrate automatic attendance log uploading
    """
    print("\n=== Attendance Uploader Demo ===")
    
    device_ip = "192.168.1.100"  # Replace with your device IP
    device = BiometricDeviceMiddleware(device_ip)
    
    try:
        # Configure attendance log uploader
        target_uri = "http://your-server.com/attendance"  # Replace with your server
        interval = 300  # 5 minutes
        
        print(f"Configuring attendance uploader to {target_uri} every {interval} seconds...")
        device.config_attend_log_uploader(target_uri, interval)
        
        # Check uploader status
        status = device.get_attend_log_uploader_status()
        print(f"Uploader status: {status}")
        
    except BiometricDeviceError as e:
        print(f"Error configuring uploader: {e}")


if __name__ == "__main__":
    # Run main demo
    main()
    
    # Uncomment to run additional demos
    # demo_photo_to_face()
    # demo_attendance_uploader()

File: example-usage.js
#!/usr/bin/env node
/**
 * Biometric Device Middleware - Node.js Example Usage
 * 
 * This script demonstrates how to use the BiometricDeviceMiddleware to interact
 * with biometric devices using the specified protocol.
 */

const { BiometricDeviceMiddleware, BiometricDeviceError } = require('./biometric-middleware');

async function main() {
    // Initialize the middleware
    const deviceIp = '192.168.1.100';  // Replace with your device IP
    const apiKey = 'your_api_key';     // Replace with your API key (if configured)
    
    // Create middleware instance
    const device = new BiometricDeviceMiddleware(deviceIp, apiKey);
    
    console.log('=== Biometric Device Middleware Demo ===\n');
    
    try {
        // 1. Get device information
        console.log('1. Getting device information...');
        const deviceInfo = await device.getVersionInfo();
        console.log(`   Firmware version: ${deviceInfo.firmware_version}`);
        console.log(`   Face algorithm: ${deviceInfo.face_algorithm_version || 'N/A'}`);
        console.log(`   Fingerprint algorithm: ${deviceInfo.fp_algorithm_version || 'N/A'}`);
        
        // Get device capabilities
        const capabilities = await device.getDeviceCapabilities();
        console.log(`   Capabilities:`, capabilities);
        
        // Get device usage
        const usage = await device.getCurrentUsage();
        console.log(`   Current usage:`, usage);
        
        console.log();
        
        // 2. Time synchronization
        console.log('2. Time synchronization...');
        const currentTime = await device.getDeviceTime();
        console.log(`   Current device time: ${currentTime}`);
        
        // Set device time to current system time
        await device.setDeviceTime();
        const newTime = await device.getDeviceTime();
        console.log(`   Updated device time: ${newTime}`);
        console.log();
        
        // 3. User management
        console.log('3. User management...');
        
        // Get all user IDs
        const userIds = await device.getAllUserIds();
        console.log(`   Found ${userIds.length} users: ${userIds.slice(0, 10)}`);  // Show first 10
        
        // Create a test user
        const testUserId = '999';
        console.log(`   Creating test user with ID: ${testUserId}`);
        await device.setUserInfo({
            id: testUserId,
            name: 'Test User',
            department: 'Engineering',
            privilege: 'user',
            password: '12345'
        });
        
        // Get user information
        const userInfo = await device.getUserInfo(testUserId);
        console.log(`   Test user info:`, userInfo);
        console.log();
        
        // 4. Device settings
        console.log('4. Device settings...');
        
        // Get and display current volume
        const volume = await device.getSoundVolume();
        console.log(`   Current sound volume: ${volume}`);
        
        // Get verify mode
        const verifyMode = await device.getVerifyMode();
        console.log(`   Current verify mode: ${verifyMode}`);
        
        console.log();
        
        // 5. Biometric enrollment demo (face)
        console.log('5. Biometric enrollment demo...');
        
        // Lock device before enrollment to prevent interference
        console.log('   Locking device for enrollment...');
        await device.lockDevice(true);
        
        try {
            // Start face enrollment
            console.log('   Starting face enrollment...');
            console.log('   Please look at the device camera when prompted...');
            
            const jobId = await device.beginEnrollFace();
            console.log(`   Enrollment job started with ID: ${jobId}`);
            
            // Wait for enrollment completion (with timeout)
            console.log('   Waiting for enrollment completion...');
            try {
                const result = await device.waitForEnrollmentCompletion(jobId, 30000);
                console.log(`   Enrollment successful! Face data length: ${result.face_data?.length || 0}`);
                
                // Update user with face data
                await device.setUserInfo({
                    id: testUserId,
                    faceData: result.face_data
                });
                console.log('   User updated with face data');
                
            } catch (enrollmentError) {
                console.log(`   Enrollment failed or timed out: ${enrollmentError.message}`);
            }
            
        } finally {
            // Always unlock device
            await device.lockDevice(false);
            console.log('   Device unlocked');
        }
        
        console.log();
        
        // 6. Attendance log management
        console.log('6. Attendance log management...');
        
        // Get recent attendance logs
        const logs = await device.getAttendLog();
        console.log(`   Retrieved ${logs.logs?.length || 0} attendance records`);
        
        // Show first few logs
        const logEntries = logs.logs || [];
        for (let i = 0; i < Math.min(3, logEntries.length); i++) {
            const log = logEntries[i];
            console.log(`   Log ${i + 1}: User ${log.user_id} at ${log.time} via ${log.mode}`);
        }
        
        console.log();
        
        // 7. Network configuration info
        console.log('7. Network configuration...');
        const networkConfig = await device.getNetworkConfig();
        
        if (networkConfig.ethernet) {
            const ethConfig = networkConfig.ethernet;
            const running = ethConfig.running || {};
            console.log(`   Ethernet IP: ${running.address}`);
        }
        
        if (networkConfig.wlan) {
            const wlanConfig = networkConfig.wlan;
            const running = wlanConfig.running || {};
            console.log(`   WiFi IP: ${running.address}`);
        }
        
        console.log();
        
        // 8. Clean up test user
        console.log('8. Cleaning up...');
        try {
            await device.deleteUserInfo(testUserId);
            console.log(`   Test user ${testUserId} deleted`);
        } catch (deleteError) {
            console.log(`   Test user ${testUserId} not found (already deleted?)`);
        }
        
    } catch (error) {
        if (error instanceof BiometricDeviceError) {
            console.error(`Error communicating with device: ${error.message}`);
            if (error.errorCode) {
                console.error(`Error code: ${error.errorCode}`);
            }
        } else {
            console.error(`Unexpected error: ${error.message}`);
        }
    }
    
    console.log('\n=== Demo completed ===');
}

async function demoPhotoToFace() {
    /**
     * Demonstrate photo to face data conversion
     */
    console.log('\n=== Photo to Face Data Demo ===');
    
    const deviceIp = '192.168.1.100';  // Replace with your device IP
    const device = new BiometricDeviceMiddleware(deviceIp);
    
    // This is a placeholder - you would load an actual base64-encoded JPG image
    // const photoBase64 = await loadPhotoAsBase64('path/to/photo.jpg');
    
    console.log('Note: This demo requires a base64-encoded JPG photo.');
    console.log('To use this feature, uncomment and modify the code above.');
    
    // Example usage:
    // try {
    //     const result = await device.photoToFaceData(photoBase64);
    //     if (result.state === 'succeeded') {
    //         const faceData = result.face_data;
    //         console.log(`Face data extracted successfully! Length: ${faceData.length}`);
    //         // You can now use this faceData with setUserInfo
    //     } else {
    //         console.log(`Photo conversion failed: ${result.state}`);
    //     }
    // } catch (error) {
    //     console.error(`Error: ${error.message}`);
    // }
}

async function demoAttendanceUploader() {
    /**
     * Demonstrate automatic attendance log uploading
     */
    console.log('\n=== Attendance Uploader Demo ===');
    
    const deviceIp = '192.168.1.100';  // Replace with your device IP
    const device = new BiometricDeviceMiddleware(deviceIp);
    
    try {
        // Configure attendance log uploader
        const targetUri = 'http://your-server.com/attendance';  // Replace with your server
        const interval = 300;  // 5 minutes
        
        console.log(`Configuring attendance uploader to ${targetUri} every ${interval} seconds...`);
        await device.configAttendLogUploader(targetUri, interval);
        
        // Check uploader status
        const status = await device.getAttendLogUploaderStatus();
        console.log('Uploader status:', status);
        
    } catch (error) {
        console.error(`Error configuring uploader: ${error.message}`);
    }
}

async function demoAdvancedUserManagement() {
    /**
     * Demonstrate advanced user management features
     */
    console.log('\n=== Advanced User Management Demo ===');
    
    const deviceIp = '192.168.1.100';
    const device = new BiometricDeviceMiddleware(deviceIp);
    
    try {
        // Get device capacity limits
        const limits = await device.getCapacityLimit();
        console.log('Device capacity limits:', limits);
        
        // Get current usage
        const usage = await device.getCurrentUsage();
        console.log('Current usage:', usage);
        
        // Calculate remaining capacity
        console.log('\nRemaining capacity:');
        if (limits.max_user && usage.user_count) {
            console.log(`   Users: ${limits.max_user - usage.user_count} remaining`);
        }
        if (limits.max_face && usage.face_count) {
            console.log(`   Faces: ${limits.max_face - usage.face_count} remaining`);
        }
        if (limits.max_fp && usage.fp_count) {
            console.log(`   Fingerprints: ${limits.max_fp - usage.fp_count} remaining`);
        }
        
        // Demonstrate batch user operations
        console.log('\nBatch user operations...');
        const userIds = await device.getAllUserIds();
        console.log(`Processing ${userIds.length} users...`);
        
        // You could process users in batches here
        for (let i = 0; i < Math.min(3, userIds.length); i++) {
            const userId = userIds[i];
            const userInfo = await device.getUserInfo(userId);
            console.log(`   User ${userId}: ${userInfo.name} (${userInfo.privilege})`);
        }
        
    } catch (error) {
        console.error(`Error in advanced user management demo: ${error.message}`);
    }
}

// Helper function to load photo as base64 (example implementation)
async function loadPhotoAsBase64(imagePath) {
    const fs = require('fs').promises;
    try {
        const imageBuffer = await fs.readFile(imagePath);
        return imageBuffer.toString('base64');
    } catch (error) {
        throw new Error(`Failed to load image: ${error.message}`);
    }
}

// Main execution
if (require.main === module) {
    // Run main demo
    main()
        .then(() => {
            console.log('\nAll demos completed successfully!');
        })
        .catch((error) => {
            console.error('Demo failed:', error.message);
            process.exit(1);
        });
    
    // Uncomment to run additional demos
    // demoPhotoToFace();
    // demoAttendanceUploader();
    // demoAdvancedUserManagement();
}

module.exports = {
    main,
    demoPhotoToFace,
    demoAttendanceUploader,
    demoAdvancedUserManagement,
    loadPhotoAsBase64
};
